lib/tasks.js

```diff
@@ -0,0 +1,397 @@
+            if (matches.length === 0) {
+              throw new Error(`glob "${task}" does not match any registered tasks`);
+            }
+            tasks.push.apply(tasks, matches);
+            continue;
+          }
+
+          tasks.push(task);
+          continue;
+        }
+
+        throw new TypeError('expected task dependency to be a string or function');
+      }
+      return tasks;
+    }
+
+    /**
+     * Run one or more tasks.
+     *
+     * ```js
+     * const build = app.series(['foo', 'bar', 'baz']);
+     * // promise
+     * build().then(console.log).catch(console.error);
+     * // or callback
+     * build(function() {
+     *   if (err) return console.error(err);
+     * });
+     * ```
+     * @name .build
+     * @param {object|array|string|function} `tasks` One or more tasks to run, options, or callback function. If no tasks are defined, the default task is automatically run.
+     * @param {function} `callback` (optional)
+     * @return {undefined}
+     * @api public
+     */
+
+    build(...args) {
+      const state = { status: 'starting', time: new Timer(), app: this };
+      state.time.start();
+      this.emit('build', state);
+
+      args = utils.flatten(args);
+      const cb = typeof utils.last(args) === 'function' ? args.pop() : null;
+
+      let { options, tasks } = createOptions(this, ...args);
+      if (!tasks.length) tasks = ['default'];
+
+      const each = options.parallel ? this.parallel : this.series;
+      const build = each.call(this, options, ...tasks);
+      const promise = build()
+        .then(() => {
+          state.time.end();
+          state.status = 'finished';
+          this.emit('build', state);
+        });
+
+      return resolveBuild(this, promise, cb);
+    }
+
+    /**
+     * Compose a function to run the given tasks in series.
+     *
+     * ```js
+     * const build = app.series(['foo', 'bar', 'baz']);
+     * // promise
+     * build().then(console.log).catch(console.error);
+     * // or callback
+     * build(function() {
+     *   if (err) return console.error(err);
+     * });
+     * ```
+     * @name .series
+     * @param {object|array|string|function} `tasks` Tasks to run, options, or callback function. If no tasks are defined, the `default` task is automatically run, if one exists.
+     * @param {function} `callback` (optional)
+     * @return {promise|undefined} Returns a promise if no callback is passed.
+     * @api public
+     */
+
+    series(...args) {
+      const stack = new Set();
+      const compose = this.iterator(async(tasks, options, resolve) => {
+        for (const ele of tasks) {
+          const task = this.getTask(ele);
+          task.series = true;
+
+          if (task.skip(options) || stack.has(task)) {
+            continue;
+          }
+
+          const run = task.run(options);
+          task.on('finished', () => stack.delete(task));
+          task.on('starting', () => stack.add(task));
```

Comment:
Should these be `task.once`? Otherwise, if a task is executed multiple times (like when using a watch task) I think it will cause a memory leak since `stack` will never be released from memory.